### PR TITLE
Revampt download to local dir process

### PR DESCRIPTION
**Pull request by Wauplin**
Implements https://github.com/huggingface/huggingface_hub/issues/1738 (and especially https://github.com/huggingface/huggingface_hub/issues/1738#issuecomment-1907883068) :see_no_evil: 

## What does this PR do?

- Downloading files with `local_dir` do not use the cache but rely on a `.huggingface/` folder instead
- Prevent `.huggingface/` from being committed 
- Adapt CLI
- Revampt all "local dir" tests
- All downloads (to local dir or to cache) will not resume by default any incomplete downloads. This should prove useful on slow connections. IMO there is very little benefit to make this optional. Only counter-example: if `hf_transfer` is enabled, we do not resume download (not supported). One can use `force_download` to force a download from scratch.

## How it works?

When downloading a file `file.txt` to local dir `data/`:
1. Checks if:
    1.  `data/file.txt` exists
    2. and `data/.huggingface/download/file.txt.metadata` exists
    3. and `data/file.txt` has been modified before the metadata file was saved (metadata contains a timestamp)
2. If (1.) is true, then we consider the metadata as valid. Metadata contains `commit_hash` and `etag`. Otherwise we consider that we don't have any info on the local file.
3. if `revision == metadata.commit_hash`, then the file is valid => return
4. otherwise, we fetch remote metadata
    1. If `remote etag == metadata.etag` => update local metadata => return
    2. if `remote tag` is a sha256 and we don't have local metadata => we hash local file. If sha256 == remote etag => it's a valid LFS file => return
    3. Otherwise (etag mismatch and not LFS or LFS but sha256 mismatch) => download file

If `force_download=True` is passed, all of the above is skipped => we download the file no matter what.

## What to review?

This is a large PR (`+1,265 −760`) as it touches in depth the download logic. However a lot of the changes are about moving parts of code into private helpers to avoid duplicating the logic between `_hf_hub_download_local_dir` and `_hf_hub_download_cache_dir`.

**Important changes are:**
- logic inside [`_local_folder.py`](https://github.com/huggingface/huggingface_hub/blob/1738-revampt-download-local-dir/src/huggingface_hub/_local_folder.py) => handles metadata in the local folder (i.e. inside `./huggingface/`)
- [`file_download.py`](https://github.com/huggingface/huggingface_hub/blob/1738-revampt-download-local-dir/src/huggingface_hub/file_download.py) => where everything happens. Best to read the file instead of raw changes. Most important part is `_hf_hub_download_local_dir` while `hf_hub_download` and `_hf_hub_download_cache_dir` are iso-feature compared to before.
- [`test_file_download.py`](https://github.com/huggingface/huggingface_hub/blob/1738-revampt-download-local-dir/tests/test_file_download.py#L751): all the new test cases in `HfHubDownloadToLocalDir`

**Doc changes**:
- `cli.md`, `download.md`, `environment_variables.md`
- `snapshot_download.py` => only some docs + few tweaks, no real update
- `hf_api.py` => only some docs + few tweaks, no real update

**Less important:**
- logic to prevent from committing `huggingface/` folder (`_commit_api.py` + `test_commit_api.py` + `test_utils_paths.py`)
- `test_cli.py` => not relevant
- `command/download.py` => deprecated `--local-dir-use-symlinks` in CLI
- `constants.py`/`hub_mixin.py`/`keras_mixin.py` => some deprecation

## Example

Download `README.md` and `model.safetensors` from `gpt2` repo into `./data/gpt2` folder:

```sh
huggingface-cli download gpt2 README.md model.safetensors --local-dir=data/gpt2
```

Resulting tree:

```sh
# tree -alh data/
[4.0K]  data/
└── [4.0K]  gpt2
    ├── [4.0K]  .huggingface
    │   ├── [4.0K]  download
    │   │   ├── [   0]  model.safetensors.lock
    │   │   ├── [ 182]  model.safetensors.metadata
    │   │   ├── [   0]  README.md.lock
    │   │   └── [ 158]  README.md.metadata
    │   └── [   1]  .gitignore
    ├── [523M]  model.safetensors
    └── [7.9K]  README.md

3 directories, 7 files
```

## How to try it?

```
pip install git+https://github.com/huggingface/huggingface_hub@1738-revampt-download-local-dir
```

## TODO

- [ ] add an explanation guide on "how `.huggingface/` works?" (similar to the "Cache" guide?)
- [ ] extensively test corner cases

### CODE REVIEW

File: src/huggingface_hub/commands/download.py [hunk 1]
```diff
@@ -94,13 +94,7 @@ def register_subcommand(parser: _SubParsersAction):
...
             "--local-dir-use-symlinks",
             choices=["auto", "True", "False"],
-            default="auto",
-            help=(
-                "To be used with `local_dir`. If set to 'auto', the cache directory will be used and the file will be"
-                " either duplicated or symlinked to the local directory depending on its size. It set to `True`, a"
-                " symlink will be created, no matter the file size. If set to `False`, the file will either be"
-                " duplicated from cache (if already exists) or downloaded from the Hub and not cached."
-            ),
+            help=("Deprecated and ignored. Downloading to a local directory do not use symlinks anymore."),
```
Comment (LysandreJik) [inline src/huggingface_hub/commands/download.py:97]:
```suggestion
            help=("Deprecated and ignored. Downloading to a local directory does not use symlinks anymore."),
```

File: src/huggingface_hub/commands/download.py [hunk 2]
```diff
@@ -134,19 +128,10 @@ def __init__(self, args: Namespace) -> None:
...
-            self.local_dir_use_symlinks = False
-        elif use_symlinks_lowercase == "auto":
-            self.local_dir_use_symlinks = "auto"
-        else:
-            raise ValueError(
-                f"'{args.local_dir_use_symlinks}' is not a valid value for `local_dir_use_symlinks`. It must be either"
-                " 'auto', 'True' or 'False'."
+        if args.local_dir_use_symlinks is not None:
+            warnings.warn(
+                "Ignoring --local-dir-use-symlinks. Downloading to a local directory do not use symlinks anymore.",
```
Comment (LysandreJik) [inline src/huggingface_hub/commands/download.py:133]:
```suggestion
                "Ignoring --local-dir-use-symlinks. Downloading to a local directory does not use symlinks anymore.",
```

File: src/huggingface_hub/file_download.py [hunk 3]
```diff
@@ -671,6 +672,13 @@ def cached_download(
             " 'hf_hub_download'",
             FutureWarning,
         )
+    if resume_download:
+        warnings.warn(
+            "`resume_download` is deprecated and will be removed in version 1.0.0. "
+            "Download always resume when possible. "
```
Comment (LysandreJik) [inline src/huggingface_hub/file_download.py:678]:
```suggestion
            "Downloads always resume when possible. "
```

Comment (LysandreJik) [reply src/huggingface_hub/file_download.py:678]: How about passing the default of `resume_download` to `None` and having the following warning additionally:

```py
if not resume_download and resume_download is not None:
    warnings.warn("resume_download is deprecated and will be removed in 1.0.0. In order to force a download, please use force_download=True")
```

Comment (LysandreJik) [reply src/huggingface_hub/file_download.py:678]: (if a user was passing this arg with `False` they wouldn't be aware of that change/the recommended argument)

File: src/huggingface_hub/_snapshot_download.py [hunk 4]
```diff
@@ -63,20 +64,10 @@ def snapshot_download(
...
-      - If `local_dir_use_symlinks=True`, files are downloaded, stored in the cache directory and symlinked in `local_dir`.
-        This is optimal in term of disk usage but files must not be manually edited.
-      - If `local_dir_use_symlinks=False` and the blob files exist in the cache directory, they are duplicated in the
-        local dir. This means disk usage is not optimized.
-      - Finally, if `local_dir_use_symlinks=False` and the blob files do not exist in the cache directory, then the
-        files are downloaded and directly placed under `local_dir`. This means if you need to download them again later,
-        they will be re-downloaded entirely.
+    If `local_dir` is provided, the file structure from the repo will be replicated in this location. When using this
+    option, the `cache_dir` will not be used and a `.huggingface/` folder will be created at the root of `local_dir`
+    to store some metadata related to the downloaded files.While this mechanism is not as robust as the main
```
Comment (LysandreJik) [inline src/huggingface_hub/_snapshot_download.py:69]:
```suggestion
    to store some metadata related to the downloaded files. While this mechanism is not as robust as the main
```

File: docs/source/en/package_reference/environment_variables.md [hunk 5]
```diff
@@ -67,10 +67,7 @@ For more details, see [logging reference](../package_reference/utilities#hugging
 
 ### HF_HUB_LOCAL_DIR_AUTO_SYMLINK_THRESHOLD
 
-Integer value to define under which size a file is considered as "small". When downloading files to a local directory,
-small files will be duplicated to ease user experience while bigger files are symlinked to save disk usage.
-
-For more details, see the [download guide](../guides/download#download-files-to-local-folder).
+This environment variable has been deprecated and is now ignored by `huggingface_hub`. Downloading files to the local dir do not rely on symlinks anymore.
```
Comment (LysandreJik) [inline docs/source/en/package_reference/environment_variables.md:70]:
```suggestion
This environment variable has been deprecated and is now ignored by `huggingface_hub`. Downloading files to the local dir does not rely on symlinks anymore.
```

**Review by LysandreJik — APPROVED** [review]
Awesome! Super nice side effect of having `resume_downloads` on by default.
I played with it locally and we already discussed potential improvements. 

The rest LGTM!

File: src/huggingface_hub/file_download.py (hunk 3 above)
Comment (Wauplin) [reply src/huggingface_hub/file_download.py]: Good idea yes! I've updated this in https://github.com/huggingface/huggingface_hub/pull/2223/commits/92106486701ef8175920b0bade7bf84f5e3c010e

File: docs/source/en/guides/download.md [hunk 6]
```diff
@@ -126,42 +126,25 @@ files except `vocab.json`.
...
+## Download file(s) to a local folder
+
+By default, we recommend using the [cache system](./manage-cache) to download files from the Hub. You can specify a custom cache location using the `cache_dir` parameter in [`hf_hub_download`] and [`snapshot_download`], or by setting the [`HF_HOME`](../package_reference/environment_variables#hf_home) environment variable.
+
+However, if you need to download files to a specific folder, you can pass a `local_dir` parameter to the download function. This is useful to get a workflow closer to what `git` commands offer. The downloaded files will maintain their original file structure within the specified folder. For example, if `filename="data/train.csv"` and `local_dir="path/to/folder"`, the resulting filepath will be `"path/to/folder/data/train.csv"`.
+
+Note that a `.huggingface/` folder will be created at the root of your local directory, containing metadata about the downloaded files. This prevents re-downloading files if you re-run your script. While this mechanism is not as robust as the main cache-system, it's optimized for regularly pulling the latest version of a repository.
```
Comment (stevhliu) [inline docs/source/en/guides/download.md:135]: Same comment about the regularly pulling part.

File: docs/source/en/guides/download.md [hunk 7]
```diff
@@ -126,42 +126,25 @@ files except `vocab.json`.
...
+
+<Tip>
```
Comment (stevhliu) [inline docs/source/en/guides/download.md:137]: I would leave this as regular text since we want the user to be aware of what happens if they delete the folder. A tip may suggest to the user that they can skip or gloss over this section which we don't necessarily want. Plus its nice to not have two `Tip`'s in a row :)

File: docs/source/en/guides/cli.md [hunk 8]
```diff
@@ -224,18 +224,20 @@ The examples above show how to download from the latest commit on the main branc
 
 ### Download to a local folder
 
-The recommended (and default) way to download files from the Hub is to use the cache-system. However, in some cases you want to download files and move them to a specific folder. This is useful to get a workflow closer to what git commands offer. You can do that using the `--local_dir` option. The file is downloaded to a tmp file and then moved to the local dir to avoid having partially downloaded files in the local folder.
+The recommended (and default) way to download files from the Hub is to use the cache-system. However, in some cases you want to download files and move them to a specific folder. This is useful to get a workflow closer to what git commands offer. You can do that using the `--local_dir` option.
 
-<Tip warning={true}>
+Note that a `.huggingface/` folder will be created at the root of your local directory, containing metadata about the downloaded files. This prevents re-downloading files if you re-run the command. While this mechanism is not as robust as the main cache-system, it's optimized for regularly pulling the latest version of a repository.
```
[the remark this answers is not in the document]
Comment (Wauplin) [reply docs/source/en/guides/cli.md:229]: Sorry if it wasn't clear. So basically:
- download to cache: optimized for any cases. Can pull multiple times latest version but also switch between revisions. No duplication involved but some symlinks are used. 
- download to local dir: optimized only for pulling latest changes. It checks in the metadata if the file has changed or not (based on remote information as well) and if yes, then it download the new version. If you try to switch between different revisions, then each time you switch the file will be redownloaded (not optimized). It's not a problem as it's not primarily meant for that.

[the remark this answers is not in the document]
Comment (Wauplin) [reply docs/source/en/guides/cli.md:229]: So the suggestion above is not correct actually. I'll try to rephrase it.

[the remark this answers is not in the document]
Comment (stevhliu) [reply docs/source/en/guides/cli.md:229]: Ah ok I see now! Maybe something like:

```md
A `./huggingface/` folder is created at the root of your local directory containing metadata about the downloaded files. This prevents re-downloading files if they're already up-to-date. If the metadata has changed, then the new file version is downloaded. This makes the `local_dir` optimized for pulling only the latest changes.
```

[the remark this answers is not in the document]
Comment (Wauplin) [reply docs/source/en/guides/cli.md]: Much better than my wording! Thanks :) Applied changes in https://github.com/huggingface/huggingface_hub/pull/2223/commits/4f6f531581c591a100ee54aad2340d75bcb45e6e

File: src/huggingface_hub/file_download.py [hunk 9]
```diff
@@ -1696,14 +1666,241 @@ def get_hf_file_metadata(
...
+    """Download content from a URL to a destination path.
+
+    Internal logic:
+    - return early if file is already downloaded
+    - resume download if possible (from incomplete file)
+    - do not resume download if `force_download=True` or `HF_HUB_ENABLE_HF_TRANSFER=True`
+    - check disk space before downloading
+    - download content to a temporary file
+    - set correct permissions on temporary file
+    - move the temporary file to the destination path
```
Comment (pcuenca) [inline src/huggingface_hub/file_download.py:1851]:
```suggestion
    - return early if file is already downloaded at `destination_path`
    - resume download if possible (from `incomplete_path`)
    - do not resume download if `force_download=True` or `HF_HUB_ENABLE_HF_TRANSFER=True`
    - check disk space before downloading
    - download content to `incomplete_path`
    - set correct permissions on `incomplete_path`
    - move the file from `incomplete_path` to `destination_path`
```

(From the name, I was initially under the impression that this function would create a temporary file, but I see it doesn't. Suggested a few nits to be more precise with the terminology :))

File: src/huggingface_hub/commands/download.py [hunk 10]
```diff
@@ -94,21 +93,17 @@ def register_subcommand(parser: _SubParsersAction):
...
         download_parser.add_argument(
             "--force-download",
             action="store_true",
             help="If True, the files will be downloaded even if they are already cached.",
         )
         download_parser.add_argument(
-            "--resume-download", action="store_true", help="If True, resume a previously interrupted download."
+            "--resume-download",
+            action="store_true",
+            help="Deprecated and ignored. Downloading a file to local dir always attempt to resume previously interrupted download.",
```
Comment (pcuenca) [inline src/huggingface_hub/commands/download.py:106]:
```suggestion
            help="Deprecated and ignored. Downloading a file to local dir always attempts to resume previously interrupted downloads (unless hf-transfer is enabled).",
```

I wonder if we should be more clear about hf-transfer, I was using `HF_HUB_ENABLE_HF_TRANSFER=1` and only realized this behaviour after looking at the pr.

**Review by pcuenca — COMMENTED** [review]
Looks great! (from a distance :))

File: src/huggingface_hub/file_download.py (hunk 9 above)
Comment (Wauplin) [reply src/huggingface_hub/file_download.py:1851]: Much clearer, thanks!

File: src/huggingface_hub/commands/download.py (hunk 10 above)
Comment (Wauplin) [reply src/huggingface_hub/commands/download.py]: That's also why I stopped to promote `hf_transfer` when running the download CLI. `hf_transfer` is more limited in term of features and in particular is not able to resume an incomplete file (by design). I might update the `hf-transfer` docs at some point to explain more the limitations but in the meantime, let's not promote it at all. (+I added your changes)

**Comment by Wauplin** [comment]
Thanks @stevhliu @pcuenca for the review! I like it better like this :D

**Comment by Wauplin** [comment]
Addressed all comments and fixed the CI (failures were not only due to this PR but I fixed them here anyway).

Let's get this merged! :tada:
